src/value/owned.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+//! Provides owned implementations of [`SymbolToken`], [`Element`] and its dependents.
+//!
+//! This API is simpler to manage with respect to borrowing lifetimes, but requires full
+//! ownership of data to do so.
+
+use super::{AnyInt, Element, ImportSource, Sequence, Struct, SymbolToken};
+use crate::types::SymbolId;
+use crate::IonType;
+
+/// An owned implementation of  [`ImportSource`].
+#[derive(Debug, Clone)]
+pub struct OwnedImportSource {
+    table: String,
+    sid: SymbolId,
+}
+
+impl OwnedImportSource {
+    pub fn new<T: Into<String>>(table: T, sid: SymbolId) -> Self {
+        Self {
+            table: table.into(),
+            sid,
+        }
+    }
+}
+
+impl ImportSource for OwnedImportSource {
+    fn table(&self) -> &str {
+        &self.table
+    }
+
+    fn sid(&self) -> usize {
+        self.sid
+    }
+}
+
+/// An owned implementation of [`SymbolToken`].
+#[derive(Debug, Clone)]
+pub struct OwnedSymbolToken {
+    text: Option<String>,
+    local_sid: Option<SymbolId>,
+    source: Option<OwnedImportSource>,
```

Comment:
Is it ok for an implementer of `SymbolToken` to always return `None` for the source? (e.g. If I want to make a `SkinnySymbolToken` type that doesn't have an extra 9 bytes+alignment, can I decide to not track the import source?)

Reply:
Yes, it is _okay_, but only if the token `text` is not `None`.  So for applications that know the text of all symbols at compile time you could statically use such an implementation or hide that detail with a `dyn SymbolToken`.